File: packages/php/src/Transport.php
<?php

declare(strict_types=1);

namespace Opensms;

use InvalidArgumentException;

/**
 * Request pipeline shared by every resource: headers, JSON encoding,
 * Idempotency-Key handling, retries with backoff (honouring Retry-After),
 * and mapping non-2xx responses to {@see OpensmsException}. The actual
 * network call is delegated to the handler ({@see CurlHandler} by default).
 *
 * @internal
 */
final class Transport
{
    public const VERSION = '0.1.0';
    public const USER_AGENT = 'opensms-php/' . self::VERSION;
    public const DEFAULT_BASE_URL = 'https://api.opensms.io';

    /** Retry-After values above this many seconds are not waited for. */
    private const MAX_RETRY_AFTER = 60.0;

    private readonly string $apiKey;
    private readonly string $baseUrl;
    private readonly float $timeout;
    private readonly int $maxRetries;
    /** @var callable(array<string, mixed>): array<string, mixed> */
    private $handler;
    /** @var callable(float): void */
    private $sleeper;

    /**
     * @param array{baseUrl?: string, timeout?: float|int, maxRetries?: int, handler?: callable, sleep?: callable} $options
     */
    public function __construct(string $apiKey, array $options = [])
    {
        $this->apiKey = $apiKey;
        $this->baseUrl = rtrim((string) ($options['baseUrl'] ?? self::DEFAULT_BASE_URL), '/');
        $this->timeout = (float) ($options['timeout'] ?? 30.0);
        $maxRetries = (int) ($options['maxRetries'] ?? 2);
        if ($maxRetries < 0) {
            throw new InvalidArgumentException('OpenSMS: `maxRetries` must be 0 or more.');
        }
        $this->maxRetries = $maxRetries;
        $this->handler = $options['handler'] ?? new CurlHandler();
        $this->sleeper = $options['sleep'] ?? static function (float $seconds): void {
            if ($seconds > 0) {
                usleep((int) round($seconds * 1_000_000));
            }
        };
    }

    public function baseUrl(): string
    {
        return $this->baseUrl;
    }

    /**
     * Perform one API call.
     *
     * @param array<string, mixed> $query Already-mapped query parameters (null values are skipped).
     * @param array<string, mixed>|null $json JSON body, or null for none.
     * @param array{
     *     idempotent?: bool,
     *     idempotencyKey?: ?string,
     *     retry?: bool,
     *     rawBody?: string,
     *     contentType?: string
     * } $opts `idempotent` sends an Idempotency-Key (the given one or a fresh
     *     UUIDv4, reused on every retry). `retry` false disables retries for
     *     a POST that is not idempotent. `rawBody` + `contentType` send a
     *     non-JSON body (CSV batches).
     */
    public function request(string $method, string $path, array $query = [], ?array $json = null, array $opts = []): mixed
    {
        $headers = [
            'Authorization' => 'Bearer ' . $this->apiKey,
            'Accept' => 'application/json',
            'User-Agent' => self::USER_AGENT,
        ];

        $body = null;
        if (isset($opts['rawBody'])) {
            $body = $opts['rawBody'];
            $headers['Content-Type'] = $opts['contentType'] ?? 'application/octet-stream';
        } elseif ($json !== null) {
            $body = json_encode($json, JSON_THROW_ON_ERROR | JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
            $headers['Content-Type'] = 'application/json';
        }

        $idempotent = (bool) ($opts['idempotent'] ?? false);
        if ($idempotent) {
            $key = $opts['idempotencyKey'] ?? null;
            $headers['Idempotency-Key'] = ($key !== null && $key !== '') ? (string) $key : self::uuid4();
        }

        // GET/PUT/PATCH/DELETE are always safe to repeat; POST only with a key.
        $retryable = ($opts['retry'] ?? true) && ($method !== 'POST' || $idempotent);
        $maxAttempts = $retryable ? $this->maxRetries + 1 : 1;

        $request = [
            'method' => $method,
            'url' => $this->baseUrl . $path . Models::queryString($query),
            'headers' => $headers,
            'body' => $body,
            'timeout' => $this->timeout,
        ];

        $lastNetworkError = null;
        for ($attempt = 1; $attempt <= $maxAttempts; $attempt++) {
            try {
                $response = ($this->handler)($request);
            } catch (\Exception $e) {
                if ($e instanceof OpensmsException) {
                    throw $e;
                }
                $lastNetworkError = $e;
                if ($attempt < $maxAttempts) {
                    ($this->sleeper)($this->backoff($attempt));
                    continue;
                }
                break;
            }

            $status = (int) $response['status'];
            $respHeaders = array_change_key_case($response['headers'] ?? [], CASE_LOWER);
            $raw = (string) ($response['body'] ?? '');

            if ($status >= 200 && $status < 300) {
                return self::decodeSuccess($status, $raw);
            }

            $retryAfter = self::parseRetryAfter($respHeaders['retry-after'] ?? null);
            if ($attempt < $maxAttempts && self::isRetryableStatus($status)
                && ($retryAfter === null || $retryAfter <= self::MAX_RETRY_AFTER)) {
                ($this->sleeper)($retryAfter ?? $this->backoff($attempt));
                continue;
            }

            throw self::toError($status, $respHeaders, $raw, $retryAfter);
        }

        $reason = $lastNetworkError?->getMessage() ?? 'unknown error';

        throw new OpensmsException(0, 'OpenSMS request failed: ' . $reason, previous: $lastNetworkError);
    }

    private static function isRetryableStatus(int $status): bool
    {
        return in_array($status, [429, 500, 502, 503, 504], true);
    }

    /** Full-jitter exponential backoff: random(0, min(8, 0.5 * 2^(n-1))). */
    private function backoff(int $attempt): float
    {
        $cap = min(8.0, 0.5 * (2 ** ($attempt - 1)));

        return $cap * (mt_rand() / mt_getrandmax());
    }

    /** Integer seconds or an HTTP date, in seconds from now; null if absent or unparseable. */
    public static function parseRetryAfter(?string $value): ?float
    {
        if ($value === null) {
            return null;
        }
        $value = trim($value);
        if ($value === '') {
            return null;
        }
        if (preg_match('/^\d+(\.\d+)?$/', $value) === 1) {
            return (float) $value;
        }
        $when = strtotime($value);
        if ($when === false) {
            return null;
        }

        return (float) max(0, $when - time());
    }

    private static function decodeSuccess(int $status, string $raw): mixed
    {
        if ($status === 204 || $raw === '') {
            return null;
        }
        try {
            return json_decode($raw, true, 512, JSON_THROW_ON_ERROR | JSON_BIGINT_AS_STRING);
        } catch (\JsonException) {
            return $raw;
        }
    }

    /**
     * @param array<string, string> $headers lowercase header names
     */
    private static function toError(int $status, array $headers, string $raw, ?float $retryAfter): OpensmsException
    {
        $decoded = null;
        if ($raw !== '') {
            try {
                $decoded = json_decode($raw, true, 512, JSON_THROW_ON_ERROR);
            } catch (\JsonException) {
                $decoded = null;
            }
        }

        $problem = is_array($decoded) && !array_is_list($decoded) ? $decoded : null;
        $str = static fn (string $k): ?string => ($problem !== null && isset($problem[$k]) && is_string($problem[$k])) ? $problem[$k] : null;

        $detail = $str('detail');
        $title = $str('title');
        $errors = ($problem !== null && isset($problem['errors']) && is_array($problem['errors'])) ? $problem['errors'] : null;
        $requestId = $headers['x-request-id'] ?? null;

        $message = $detail ?? $title ?? "OpenSMS request failed with status {$status}";

        return new OpensmsException(
            status: $status,
            message: $message,
            type: $str('type'),
            title: $title,
            detail: $detail,
            errorCode: $str('code'),
            traceId: $str('trace_id'),
            errors: $errors,
            requestId: ($requestId !== null && $requestId !== '') ? $requestId : null,
            retryAfter: $retryAfter,
            body: $decoded ?? ($raw === '' ? null : $raw),
        );
    }

    /** RFC 4122 version 4 UUID (36 characters). */
    public static function uuid4(): string
    {
        $bytes = random_bytes(16);
        $bytes[6] = chr((ord($bytes[6]) & 0x0f) | 0x40);
        $bytes[8] = chr((ord($bytes[8]) & 0x3f) | 0x80);
        $hex = bin2hex($bytes);

        return sprintf(
            '%s-%s-%s-%s-%s',
            substr($hex, 0, 8),
            substr($hex, 8, 4),
            substr($hex, 12, 4),
            substr($hex, 16, 4),
            substr($hex, 20, 12),
        );
    }
}

File: packages/go/transport.go
package opensms

import (
	"bytes"
	"context"
	"crypto/rand"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	mrand "math/rand"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

const (
	// Version is the SDK version, sent in the User-Agent header.
	Version = "0.1.0"

	// DefaultBaseURL is the production API origin.
	DefaultBaseURL = "https://api.opensms.io"

	defaultMaxRetries = 2
	defaultTimeout    = 30 * time.Second
	userAgent         = "opensms-go/" + Version

	maxRetryAfter = 60 * time.Second
	backoffBase   = 500 * time.Millisecond
	backoffCap    = 8 * time.Second
)

// transport is the only code that talks to the network. It owns bearer
// authentication, JSON encoding and decoding, Idempotency-Key handling,
// retries with backoff (honouring Retry-After) and error mapping to *Error.
type transport struct {
	apiKey     string
	baseURL    string
	maxRetries int
	timeout    time.Duration
	httpClient *http.Client

	// sleep and jitter are replaced in tests.
	sleep  func(ctx context.Context, d time.Duration) error
	jitter func(max time.Duration) time.Duration
}

// request describes one API call. Resources build it; transport sends it.
type request struct {
	method string
	path   string
	query  url.Values

	// body is JSON-encoded when non-nil. rawBody with contentType is sent
	// verbatim instead (CSV, multipart).
	body        any
	rawBody     []byte
	contentType string

	// idempotent marks methods that send an Idempotency-Key ("req"/"opt" in
	// SURFACE.md). The key is generated once per call and reused on retries.
	idempotent bool
	// noRetry disables automatic retries (non-idempotent POSTs).
	noRetry bool

	opts callOptions
}

// CallOption configures a single method call.
type CallOption func(*callOptions)

type callOptions struct {
	idempotencyKey string
}

// WithIdempotencyKey sets the Idempotency-Key for a call. Without it, methods
// that support idempotency send a generated UUIDv4. The same key is reused on
// every retry of that call. Ignored by methods that do not send the header.
func WithIdempotencyKey(key string) CallOption {
	return func(o *callOptions) { o.idempotencyKey = key }
}

func applyCallOptions(opts []CallOption) callOptions {
	var o callOptions
	for _, opt := range opts {
		if opt != nil {
			opt(&o)
		}
	}
	return o
}

// do sends req and decodes a 2xx JSON body into out (which may be nil).
func (t *transport) do(ctx context.Context, req request, out any) error {
	var payload []byte
	contentType := req.contentType
	if req.body != nil {
		encoded, err := json.Marshal(req.body)
		if err != nil {
			return newLocalError("", fmt.Sprintf("failed to encode request body: %v", err), err)
		}
		payload = encoded
		contentType = "application/json"
	} else if req.rawBody != nil {
		payload = req.rawBody
	}

	target := t.baseURL + req.path
	if len(req.query) > 0 {
		target += "?" + encodeQuery(req.query)
	}

	idemKey := ""
	if req.idempotent {
		idemKey = req.opts.idempotencyKey
		if idemKey == "" {
			idemKey = newUUID()
		}
	}
	retryable := !req.noRetry && (req.method != http.MethodPost || idemKey != "")

	attempts := 1
	if retryable {
		attempts += t.maxRetries
	}

	var lastErr error
	for attempt := 1; attempt <= attempts; attempt++ {
		status, header, data, err := t.send(ctx, req.method, target, payload, contentType, idemKey)
		last := attempt == attempts
		if err != nil {
			if ctx.Err() != nil {
				return newLocalError("", fmt.Sprintf("request cancelled: %v", ctx.Err()), ctx.Err())
			}
			lastErr = err
			if last {
				break
			}
			if werr := t.sleep(ctx, t.jitter(backoffDelay(attempt))); werr != nil {
				return werr
			}
			continue
		}

		if status >= 200 && status < 300 {
			if out == nil || len(bytes.TrimSpace(data)) == 0 || status == http.StatusNoContent {
				return nil
			}
			if err := json.Unmarshal(data, out); err != nil {
				return &Error{Status: status, Body: data, message: fmt.Sprintf("failed to decode response: %v", err), cause: err}
			}
			return nil
		}

		retryAfter, hasRetryAfter := parseRetryAfter(header.Get("Retry-After"), time.Now())
		apiErr := newAPIError(status, data, header.Get("X-Request-ID"), retryAfter)
		if last || !isRetryableStatus(status) {
			return apiErr
		}
		wait := t.jitter(backoffDelay(attempt))
		if hasRetryAfter {
			if retryAfter > maxRetryAfter {
				return apiErr
			}
			wait = retryAfter
		}
		if werr := t.sleep(ctx, wait); werr != nil {
			return werr
		}
	}
	return newLocalError("", fmt.Sprintf("request failed: %v", lastErr), lastErr)
}

// send performs one HTTP attempt with the per-attempt timeout and returns the
// fully read response.
func (t *transport) send(ctx context.Context, method, target string, payload []byte, contentType, idemKey string) (int, http.Header, []byte, error) {
	actx := ctx
	if t.timeout > 0 {
		var cancel context.CancelFunc
		actx, cancel = context.WithTimeout(ctx, t.timeout)
		defer cancel()
	}
	var body io.Reader
	if payload != nil {
		body = bytes.NewReader(payload)
	}
	httpReq, err := http.NewRequestWithContext(actx, method, target, body)
	if err != nil {
		return 0, nil, nil, err
	}
	httpReq.Header.Set("Authorization", "Bearer "+t.apiKey)
	httpReq.Header.Set("Accept", "application/json")
	httpReq.Header.Set("User-Agent", userAgent)
	if payload != nil && contentType != "" {
		httpReq.Header.Set("Content-Type", contentType)
	}
	if idemKey != "" {
		httpReq.Header.Set("Idempotency-Key", idemKey)
	}
	resp, err := t.httpClient.Do(httpReq)
	if err != nil {
		return 0, nil, nil, err
	}
	defer resp.Body.Close()
	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return 0, nil, nil, err
	}
	return resp.StatusCode, resp.Header, data, nil
}

func isRetryableStatus(status int) bool {
	switch status {
	case 429, 500, 502, 503, 504:
		return true
	}
	return false
}

// backoffDelay is the upper bound for retry n (1-based):
// min(8s, 0.5s * 2^(n-1)). The actual wait is a full-jitter random value in
// [0, bound].
func backoffDelay(n int) time.Duration {
	d := backoffBase
	for i := 1; i < n && d < backoffCap; i++ {
		d *= 2
	}
	if d > backoffCap {
		d = backoffCap
	}
	return d
}

func fullJitter(max time.Duration) time.Duration {
	if max <= 0 {
		return 0
	}
	return time.Duration(mrand.Int63n(int64(max) + 1))
}

func sleepContext(ctx context.Context, d time.Duration) error {
	if d <= 0 {
		return nil
	}
	timer := time.NewTimer(d)
	defer timer.Stop()
	select {
	case <-ctx.Done():
		return newLocalError("", fmt.Sprintf("request cancelled: %v", ctx.Err()), ctx.Err())
	case <-timer.C:
		return nil
	}
}

// parseRetryAfter reads integer seconds or an HTTP date.
func parseRetryAfter(v string, now time.Time) (time.Duration, bool) {
	v = strings.TrimSpace(v)
	if v == "" {
		return 0, false
	}
	if secs, err := strconv.Atoi(v); err == nil {
		if secs < 0 {
			secs = 0
		}
		return time.Duration(secs) * time.Second, true
	}
	if at, err := http.ParseTime(v); err == nil {
		d := at.Sub(now)
		if d < 0 {
			d = 0
		}
		return d, true
	}
	return 0, false
}

// encodeQuery encodes query parameters in key order. Commas in values are
// kept literal so list parameters read countries=KE,NG.
func encodeQuery(q url.Values) string {
	return strings.ReplaceAll(q.Encode(), "%2C", ",")
}

// newUUID returns a random RFC 4122 version 4 UUID.
func newUUID() string {
	var b [16]byte
	if _, err := rand.Read(b[:]); err != nil {
		panic(errors.New("opensms: crypto/rand failed: " + err.Error()))
	}
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:16])
}

// escape URL-escapes one path segment.
func escape(s string) string { return url.PathEscape(s) }

// requireID fails fast, without a request, when a path id is empty.
func requireID(name, v string) error {
	if strings.TrimSpace(v) == "" {
		return fmt.Errorf("%w: %s must not be empty", ErrInvalidArgument, name)
	}
	return nil
}

// call sends req and decodes the response into a new T.
func call[T any](ctx context.Context, t *transport, req request) (*T, error) {
	var out T
	if err := t.do(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// callSlice sends req and decodes a bare JSON array.
func callSlice[T any](ctx context.Context, t *transport, req request) ([]T, error) {
	var out []T
	if err := t.do(ctx, req, &out); err != nil {
		return nil, err
	}
	return out, nil
}
